docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,83 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or default interface method, an error will occur during the attempt to box the instance.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `isinst` &ndash; Will always place `null` on stack.
```

Comment:
Why would `isinst` always produce `null`? It would seem to be a completely valid pattern to do the following

```c#
void M<T>(T value) {
  if (value is Span<byte> span) {
  }
}
```

This pattern works for value types today.

Reply:
I'm not sure how to make this work to be honest. I see the generated code is basically what you have below. However, a `box` instruction has language that indicates it will be boxed and that can't happen for ByRefLike types. I'd say this is invalid IL if `T` is ByRefLike and will fail during JIT.

I am interpreting this case from ECMA-335:

> If typeTok is a value type, the box instruction converts val to its boxed form. When typeTok is a 
non-nullable type (§I.8.2.4), this is done by creating a new object and copying the data from val 
into the newly allocated object. If it is a nullable type, this is done by inspecting val’s HasValue 
property; if it is false, a null reference is pushed onto the stack; otherwise, the result of boxing 
val’s Value property is pushed onto the stack. 

> If typeTok is a reference type, the box instruction does returns val unchanged as obj. 

> If typeTok is a generic parameter, the behavior of box instruction depends on the actual type at 
runtime. If this type is a value type it is boxed as above, if it is a reference type then val is not 
changed. However the type tracked by verification is always “boxed” typeTok for generic parameters, regardless of whether the actual type at runtime is a value or reference type. 

The statement "If this type is a value type it is boxed as above" is what I am focusing on.

Reply:
I see. There is huge special case in the JIT that handles this case. It is commented as "match and import common box idioms". My description is valid and follows ECMA, but the JIT special cases this sequence and effectively elides the `box` instruction.

Reply:
Basically, your concern is noted and that sequence should continue to work.

Reply:
Recorded sequences and defined plan.